src/tcpconnect/main.rs

```diff
@@ -0,0 +1,258 @@
+            data6.uid = bpf_get_current_uid_gid();
+            data6.ts_us = bpf_ktime_get_ns() / 1000;
+            bpf_probe_read(&data6.saddr, sizeof(data6.saddr),
+            skp->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+            bpf_probe_read(&data6.daddr, sizeof(data6.daddr),
+            skp->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
+            data6.dport = ntohs(dport);
+            bpf_get_current_comm(&data6.task, sizeof(data6.task));
+            ipv6_events.perf_submit(ctx, &data6, sizeof(data6));";
+    let mut bpf_text: String = include_str!("bpf.c").to_string();
+    let matches = App::new("tcpconnect")
+        .about("Trace TCP connects")
+        .long_about(
+            "examples:
+./tcpconnect           # trace all TCP connect()s
+./tcpconnect -p 181    # only trace PID 181
+./tcpconnect -P 80     # only trace port 80
+./tcpconnect -P 80,81  # only trace port 80 and 81
+./tcpconnect -u 1000   # only trace UID 1000",
+        )
+        .arg(
+            Arg::with_name("uid")
+                .short("u")
+                .long("uid")
+                .help("trace this UID only, e.g. -u 2322")
+                .value_name("UID")
+                .number_of_values(1)
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("pid")
+                .short("p")
+                .long("pid")
+                .help("trace this PID only, e.g. -p 343")
+                .value_name("PID")
+                .number_of_values(1)
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("port")
+                .short("P")
+                .long("port")
+                .help("destination ports to trace, e.g. -P 22 80 443")
+                .value_name("PORT")
+                .min_values(1)
+                .required(false)
+                .takes_value(true),
+        )
+        .get_matches();
+    // first thing is to swap code
+    bpf_text = bpf_text.replace("IPV4_CODE", ipv4_trace_code);
+    bpf_text = bpf_text.replace("IPV6_CODE", ipv6_trace_code);
+    // handle pids
+    if let Some(p) = matches.value_of("p") {
+        // trace pid
+        let pid = p
+            .parse::<u32>()
+            .expect("pid should be a non-negative integer");
+        bpf_text = {
+            let uuids = format!("if (pid != {}) {{ return 0; }}", pid);
+            bpf_text.replace("FILTER_PID", uuids.as_str())
+        };
+    }
+    // handle ports
+    if let Some(port_listings) = matches.values_of("P") {
+        let mut ports_conditions: Vec<String> = Vec::new();
+        for port in port_listings {
+            let port_u16 = port.parse::<u16>().expect("port number between 1-65535");
+            let port_ntohs = ntohs(port_u16);
+            let port_condition = format!("dport != {}", port_ntohs);
+            ports_conditions.push(port_condition);
+        }
+        let port_contition_statement = ports_conditions.join(" && ");
```

Comment:
What's up with the spelling for this? You could probably move the `join()` down into the `format!` below